dev/devicelab/lib/tasks/gallery.dart

```diff
@@ -58,59 +65,54 @@ class GalleryTransitionTest {
   final String transitionDurationFile;
   final String driverFile;
 
-  Future<TaskResult> call() async {
-    final Device device = await devices.workingDevice;
-    await device.unlock();
-    final String deviceId = device.deviceId;
-    final Directory galleryDirectory = dir('${flutterDirectory.path}/dev/integration_tests/flutter_gallery');
-    await inDirectory<void>(galleryDirectory, () async {
-      String applicationBinaryPath;
-      if (deviceOperatingSystem == DeviceOperatingSystem.android) {
-        section('BUILDING APPLICATION');
-        await flutter(
-          'build',
-          options: <String>[
-            'apk',
-            '--no-android-gradle-daemon',
-            '--profile',
-            '-t',
-            'test_driver/$testFile.dart',
-            '--target-platform',
-            'android-arm,android-arm64',
-          ],
-        );
-        applicationBinaryPath = 'build/app/outputs/flutter-apk/app-profile.apk';
+  @override
+  List<String> getBuildArgs(DeviceOperatingSystem deviceOperatingSystem) {
+      switch (deviceOperatingSystem) {
+        case DeviceOperatingSystem.android:
+          return <String>[
+              'apk',
+              '--no-android-gradle-daemon',
+              '--profile',
+              '-t',
+              'test_driver/$testFile.dart',
+              '--target-platform',
+              'android-arm,android-arm64',
+            ];
+          default:
+           throw Exception('$deviceOperatingSystem has no build configuration');
       }
+    }
 
-      final String testDriver = driverFile ?? (semanticsEnabled
-          ? '${testFile}_with_semantics_test'
-          : '${testFile}_test');
-      section('DRIVE START');
-      await flutter('drive', options: <String>[
+  @override
+  List<String> getTestArgs(DeviceOperatingSystem deviceOperatingSystem, String deviceId) {
+    String applicationBinaryPath;
+    if (deviceOperatingSystem == DeviceOperatingSystem.android) {
+      applicationBinaryPath = 'build/app/outputs/flutter-apk/app-profile.apk';
```

Comment:
Since the apks are going to get built on a different machine, you should consider what might happen if we fail to copy a new APK or fail to delete the old one. In this case, I think there is a risk that the wrong APK would be run.

Consider adjusting the design so that each task gets a specific APK, like gallery_foo_bar_test_named_whatever.apk which is either derived from the test name or configured by LUCI

Reply:
The APKs will be in the swarming workdir, which is guaranteed to be cleared between tasks.

1. Can the gallery tests share the same APK (i.e. we build it once, and spawn the individual tests against it)?
2. Should I update the build step here to move `build/app/outputs/flutter-apk/*.apk` to the task name? Or is there a way to have flutter build do that?

Reply:
Or named with a udid unique to the build run that can be passed into the build and test steps.

Reply:
> Or named with a udid unique to the build run that can be passed into the build and test steps.

Grr the UI didn't update with @CaseyHillers's comment before I sent that.

> Can the gallery tests share the same APK (i.e. we build it once, and spawn the individual tests against it)?

I believe that should work if all the devices are the same architecture?   I think the default for `build apk` is to build for all valid device architectures, but for example code size analysis won't work unless a `--target-platform` is specified.  Also, as-written they are all profile mode builds, but if you decided to expand this you need to make sure a debug build isn't run on a profile test or release build on a debug test, etc. since they are all copied to the same path by default.

Reply:
Architecture won't matter if its a fat apk (which is what build apk produces by default) and we don't use any code size analysis for now. The way to do this safely would be to make the APK name a hash of all of the build input arguments - in that case as long as the name matches it would be safe to use for multiple tests.

There is no way to move the apk automatically, you'd have to copy it in the devicelab.

Reply:
I added the option to pass the application binary path. However, for migration there is still going to be the supported "default values" that are hardcoded in the tasks.

I didn't rename the build artifacts as it seems non-trivial to know which apk was used unless I explicitly add tooling to control the build command (beyond passing args). Instead, I did add a `flutter clean` step before the build to ensure no previous artifacts exist before building.